File: 剑指offer/数组/jz32.py
# -*- coding:utf-8 -*-
class Solution:
    # ¿ìÅÅ
    def quick_sort(self,arr,left,right):
        if left > right: return
        i = left
        j = right
        temp = arr[left]
        while i!=j:
            while arr[j]+temp >= temp + arr[j] and i<j:
                j -= 1
            while arr[i]+temp <= temp + arr[i] and i<j:
                i += 1
            if i<j:
                arr[i],arr[j] = arr[j],arr[i]
        arr[left],arr[i] = arr[i],arr[left]
        self.quick_sort(arr,left,i-1)
        self.quick_sort(arr,i+1,right)
        return arr
    # Ï£¶ûÅÅĞò
    def shell_sort(self,arr):
        n = len(arr)
        gap = n//2
        while gap>0:
            for i in range(gap,n):
                tmp = arr[i]
                j = i
                while j>=gap and arr[j-gap]+tmp>tmp+arr[j-gap]:
                    arr[j] = arr[j-gap]
                    j -= gap
                arr[j] = tmp
            gap = gap//2
        return arr
        
    def PrintMinNumber(self, numbers):
        # write code here
        if not numbers: return ""
        s = [str(n) for n in numbers]
        self.quick_sort(s,0,len(s)-1)
        return "".join(s)

File: 剑指offer/数组/jz32.cpp
/*
输入一个正整数数组，把数组里所有数字拼接起来排成一个数，
打印能拼接出的所有数字中最小的一个。
例如输入数组{3，32，321}，
则打印出这三个数字能排成的最小数字为321323。
*/

#include <vector>
#include <algorithm>
#include <string>

using namespace std;

class Solution {
public:
    string PrintMinNumber(vector<int> numbers) {
        vector<string> str;
        for(int i:numbers) str.push_back(to_string(i));
        sort(str.begin(), str.end(),[](string a,string b){
            return a+b<b+a;
        });
        string ans = "";
        for(string i:str) ans+=i;
        return ans;
    }
};

/*
思路：
转换成string，利用string的排序。自定义排序的方法，


lambda使用

[](string a, string b){
    return a+b<b+a;
}
*/